<?xml version="1.0"?>

<robot name="two_wheel_robot">


<link name="body_link">
  <inertial>
    <mass value="0.03" />
    <origin xyz="0 0 0.0127" />
    <inertia ixx="0.000032258" ixy="0.0" ixz="0.0"
      iyy="0.000032258" iyz="0.0"
      izz="0.000032258" />
  </inertial>

  <visual>
    <origin xyz="0 0 0.0127" />
    <geometry>
      <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_body.dae"/>
    </geometry>
  </visual>

  <collision>
    <origin xyz="0 0 0.0127" />
    <geometry>
      <!-- <box size="0.0254 0.0254 0.0254"/> -->
      <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_body.dae"/>
    </geometry>
  </collision>
</link>

<gazebo reference="body_link">
  <!-- tag name has to be mu1, though in gazebo they are mu and mu2 -->
  <mu1>0.1</mu1>
  <mu2>0.1</mu2>
</gazebo>


<!-- support the robot body from swing front and back -->
<link name="front_sphere">
  <inertial>
    <mass value="0.0000001" />
    <origin xyz="0 0 0" />
    <inertia ixx="0.00000000001" ixy="0.0" ixz="0.0"
      iyy="0.00000000001" iyz="0.0"
      izz="0.00000000001" />
  </inertial>

  <!-- make the sphere invisible -->
<!--   <visual>
    <origin xyz="0 0 0" />
    <geometry>
      <sphere radius="0.002297" />
    </geometry>
  </visual> -->

  <collision>
    <origin xyz="0 0 0" />
    <geometry>
      <sphere radius="0.002297" />
    </geometry>
  </collision>
</link>

<gazebo reference="front_sphere">
  <mu1>0.0</mu1>
  <mu2>0.0</mu2>
</gazebo>

<joint name="front_support" type="fixed">
  <parent link="body_link" />
  <child link="front_sphere" />
  <origin xyz="0.0104 0 0" rpy="0 0 0" />
</joint>


<!-- back sphere support -->
<link name="back_sphere">
  <inertial>
    <mass value="0.0000001" />
    <origin xyz="0 0 0" />
    <inertia ixx="0.00000000001" ixy="0.0" ixz="0.0"
      iyy="0.00000000001" iyz="0.0"
      izz="0.00000000001" />
  </inertial>

<!--   <visual>
    <origin xyz="0 0 0" />
    <geometry>
      <sphere radius="0.002297" />
    </geometry>
  </visual> -->

  <collision>
    <origin xyz="0 0 0" />
    <geometry>
      <sphere radius="0.002297" />
    </geometry>
  </collision>
</link>

<gazebo reference="back_sphere">
  <mu1>0.0</mu1>
  <mu2>0.0</mu2>
</gazebo>

<joint name="back_support" type="fixed">
  <parent link="body_link" />
  <child link="back_sphere" />
  <origin xyz="-0.0104 0 0" rpy="0 0 0" />
</joint>


<link name="left_wheel">
  <inertial>
    <mass value="0.002" />
    <origin xyz="0 0 0.002" />
    <inertia ixx="0.00000005659" ixy="0.0" ixz="0.0"
      iyy="0.00000005659" iyz="0.0"
      izz="0.0000001125" />
  </inertial>

  <!-- use regular cylinder if wanting smooth visual experience -->
  <!-- remember to set color to grey in gazebo tag -->
  <!-- or low definition wheel if wanting to observe wheel rotation -->
  <!-- high definition wheel is not practical, too slow to visualize -->
  <visual>
    <origin xyz="0 0 0.002" />
    <geometry>
      <cylinder radius="0.015" length="0.004" />
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_left_hd.dae"/> -->
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_left_ld.dae"/> -->
    </geometry>
  </visual>

  <!-- collision can always be simple as a cylinder -->
  <collision>
    <origin xyz="0 0 0.002" />
    <geometry>
      <cylinder radius="0.015" length="0.004" />
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_left_hd.dae"/> -->
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_left_ld.dae"/> -->
    </geometry>
  </collision>
</link>

<gazebo reference="left_wheel">
  <!-- <material>Gazebo/Grey</material> -->
  <material>Gazebo/Red</material>
  <mu1>1</mu1>
  <mu2>0.5</mu2>
</gazebo>


<!-- <link name="left_wheel_mark">
  <inertial>
    <mass value="0.0000001" />
    <origin xyz="0 0 0" />
    <inertia ixx="0.00000000001" ixy="0.0" ixz="0.0"
      iyy="0.00000000001" iyz="0.0"
      izz="0.00000000001" />
  </inertial>

  <visual>
    <origin xyz="0 0 0" />
    <geometry>
      <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_left_mark.dae"/>
    </geometry>
  </visual>

  <collision>
    <origin xyz="0 0 0" />
    <geometry>
      <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_left_mark.dae"/>
    </geometry>
  </collision>
</link>

<joint name="left_wheel_mark_fixed" type="continuous">
  <parent link="left_wheel" />
  <child link="left_wheel_mark" />
  <origin xyz="0 0 0.014" rpy="0 0 0" />
  <axis xyz="0 0 1" />
</joint> -->


<link name="right_wheel">
  <inertial>
    <mass value="0.002" />
    <origin xyz="0 0 0.002" />
    <inertia ixx="0.00000005659" ixy="0.0" ixz="0.0"
      iyy="0.00000005659" iyz="0.0"
      izz="0.0000001125" />
  </inertial>

  <visual>
    <origin xyz="0 0 0.002" />
    <geometry>
      <cylinder radius="0.015" length="0.004" />
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_right_hd.dae"/> -->
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_right_ld.dae"/> -->
    </geometry>
  </visual>

  <collision>
    <origin xyz="0 0 0.002" />
    <geometry>
      <cylinder radius="0.015" length="0.004" />
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_right_hd.dae"/> -->
      <!-- <mesh filename="package://swarm_robot_description/urdf/mesh/twowheel_right_ld.dae"/> -->
    </geometry>
  </collision>
</link>

<gazebo reference="right_wheel">
  <!-- <material>Gazebo/Grey</material> -->
  <material>Gazebo/Blue</material>
  <mu1>1</mu1>
  <mu2>0.5</mu2>
</gazebo>


<joint name="left_motor" type="continuous">
  <parent link="body_link" />
  <child link="left_wheel" />
  <origin xyz="0 0.0137 0.0127" rpy="-1.5708 0 0" />
  <axis xyz="0 0 1" />
</joint>


<joint name="right_motor" type="continuous">
  <parent link="body_link" />
  <child link="right_wheel" />
  <!-- <origin xyz="0 0.0177 0.0127" rpy="1.5708 0 0" /> -->
  <origin xyz="0 -0.0137 0.0127" rpy="1.5708 0 0" />
  <axis xyz="0 0 -1" />
<!--   axis setting, invert the direction
  so that two wheels are tuned at the same direction, which means
  when the same wheel position command is given, both wheels will make the same effect
  of pushing the robot forward or backward -->
</joint>


</robot>
